raft/node.go

```diff
@@ -114,11 +114,25 @@ type Node interface {
 	ProposeConfChange(ctx context.Context, cc pb.ConfChange) error
 	// Step advances the state machine using the given message. ctx.Err() will be returned, if any.
 	Step(ctx context.Context, msg pb.Message) error
+
 	// Ready returns a channel that returns the current point-in-time state.
-	// Users of the Node must call Advance after applying the state returned by Ready.
+	// Users of the Node must call Advance after retrieving the state returned by Ready.
+	//
+	// NOTE: No committed entries from the next Ready may be applied until all committed entries
+	// and snapshots from the previous one have finished.
 	Ready() <-chan Ready
-	// Advance notifies the Node that the application has applied and saved progress up to the last Ready.
+	// Advance notifies the Node that the application has saved progress up to the last Ready.
 	// It prepares the node to return the next available Ready.
+	//
+	// In normal case, application should call Advance after it applies the entries in last Ready.
+	// When the last Ready contains a snapshot, the application might take a long time to apply
+	// the snapshot data. To continue receiving Ready without blocking raft progress, it is safe(*)
+	// to call Advance before finish applying the last ready. But the applied index in your application
+	// will be behind the applied index that tracks by Node.
+	//
+	// (*) NOTE: If a follower changes its state to candidate, it MUST wait for finishing all previous
```

Comment:
Am I right in assuming that this is required because of configuration changes needing to be applied? 

Say I decide to queue up all non-configuration and apply them asynchronously (but apply configuration changes immediately). This is probably wrong, but if on restart I retrieve all committed entries from Storage and apply them to my state machine before starting the Node with `Applied: hardState.Commit`, will I still run into the problem?

It feels weird that there shouldn't be a way where I can simply apply changes to my state machine asynchronously, without jumping through this (seemingly tricky) hoop of checking whether my node is a candidate.